File: data_structures/python/DoublyCircularLL.py
class Node:
    def __init__(self,data):
        self.data = data
        self.prev = None
        self.next = None

class DoublyCircularLL:
    def __init__(self):
        self.head = None
        self.tail = None

    def InsertFirst(self,data):
        new_node = Node(data)
        if(self.head == None and self.tail == None):#LL is empty
            self.head = new_node
            self.tail = new_node
            new_node.prev = self.head
            new_node.next = self.tail

        else:
            new_node.next = self.head
            new_node.prev = self.tail
            self.head.prev = new_node  
            self.tail.next = new_node
            self.head = new_node

    def InsertLast(self,data):
        new_node = Node(data)
        if(self.head == None and self.tail == None):#LL is empty
            self.head = new_node
            self.tail = new_node
            new_node.prev = self.head
            new_node.next = self.tail

        else: 
            new_node.prev = self.tail
            new_node.next = self.head
            self.tail.next = new_node
            self.head.prev = new_node
            self.tail = new_node
            
    def DeleteFirst(self):
        if(self.head == None and self.tail == None):#LL is empty
            print("Linked List is empty")
            return
        
        if(self.head == self.tail):#LL has only one node
            self.head = None
            self.tail = None

        else:
            self.head = self.head.next
            self.head.prev = self.tail
            self.tail.next = self.head

    def DeleteLast(self):
        if(self.head == None and self.tail == None):#LL is empty
            print("Linked List is empty")
            return
        
        if(self.head == self.tail):#LL has only one node
            self.head = None
            self.tail = None

        else:
            temp = self.head
            while(temp.next != self.tail):
                temp = temp.next
            self.tail = temp
            self.tail.next = self.head
            self.head.prev = temp

    def Display(self):
        if(self.head == None and self.tail == None):
            print("Linked List is empty")
            return
        temp = self.head
        print("<=>",end = "")
        while True:
            print(f"|{temp.data}|<=>",end = "")
            if temp == self.tail:
                break
            temp = temp.next
        print()

    def Count(self):
        iCount = 0
        if(self.head == None and self.tail == None):
            print("Linked List is empty")
            return
        temp = self.head
        while True:
            iCount  = iCount + 1
            if temp == self.tail:
                break
            temp = temp.next
        return iCount
    
    def InsertAtPos(self,data,pos):
        iCount = self.Count()
        if(pos < 1 or pos > iCount + 1):
            print("Invalid position")
            return
        
        if(pos == 1):
            self.InsertFirst(data)

        elif(pos == iCount + 1):
            self.InsertLast(data)

        else:
            new_node = Node(data)
            temp = self.head
            for i in range(1,pos - 1):
                temp = temp.next
            new_node.next = temp.next
            new_node.prev = temp
            temp.next.prev = new_node
            temp.next = new_node
            
    def DeleteAtPos(self,pos):
        iCount = self.Count()
        if(pos < 1 or pos > iCount):
            print("Invalid position")
            return
        
        if(pos == 1):
            self.DeleteFirst()

        elif(pos == iCount):
            self.DeleteLast()
            
        else:
            temp = self.head
            for i in range(1,pos-1):
                temp = temp.next
            target = temp.next
            temp.next = target.next
            target.next.prev = temp

def main():
    sobj = DoublyCircularLL()

    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    sobj.InsertLast(101)
    sobj.InsertLast(111)
    sobj.InsertLast(121)

    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ",iRet) 

    sobj.InsertAtPos(105,5)
    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ",iRet) 

    sobj.DeleteAtPos(5)
    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ",iRet) 

if __name__ == "__main__":
    main()        
        

File: data_structures/java/DoublyCircularLL.java
class node
{
    public int data;
    public node next;
    public node prev;
}

class DoublyCircularLL
{
    private node first;
    private node last;
    private int iCount;

    public DoublyCircularLL()
    {
        first = null;
        last = null;
        iCount = 0;
    }

    public void InsertFirst(int no)
    {   
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(first == null && last == null)
        {
            first = newn;
            last = newn;

            last.next = first;
            first.prev = last;
        }
        else
        {
            newn.next = first;
            first.prev = newn;

            first = newn;

            first.prev = last;
            last.next = first;
        }
        iCount++;
    }

    public void InsertLast(int no)
    {
        node newn = null;
        newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;
        
        if(first == null && last == null)
        {
            first = newn;
            last = newn;

            last.next = first;
            first.prev = last;
        }
        else
        {
            newn.prev = last;
            last.next = newn;
            last = newn;

            last.next = first;
            first.prev = last;
        }
        iCount++;
    }

    public void Display()
    {
        if(first == null && last == null)
        {
            System.out.println("Empty Linked list");
            return;
        }
        node temp = first;
        System.out.print(" <=> ");
        do
        {
            System.out.print("| "+ temp.data + " | <=> ");
            temp = temp.next;
        }
        while(temp != first);
        System.out.println();
    }

    public int Count()
    {
        return iCount;
    }


    public void InsertAtPos(int no,int pos)
    {
        if(pos < 1 || pos > iCount + 1)
        {
            System.out.println("Invalid position.");
        }
        else if( pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            int iCnt = 0;

            node newn = null;
            newn = new node();

            newn.data = no;
            newn.next = null;
            newn.prev = null;

            node temp = first;

            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next.prev = newn;

            temp.next = newn;
            newn.prev = temp;

            iCount++;

        }
    }

    public void DeleteFirst()
    {
        if(first == null && last == null)
        {
            System.out.println("Empty linked list");
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            first = first.next;

            first.prev = last;
            last.next = first;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        if(first == null && last == null)
        {
            System.out.println("Empty linked list");
            return;
        }
        else if(first == last)
        {
            first = null;
            last = null;
        }
        else
        {
            last = last.prev;

            last.next = first;
            first.prev = last;
        }
        iCount--;
    }

    public void DeleteAtPos(int pos)
    {
        if(pos < 1 || pos > iCount)
        {
            System.out.println("Invalid position.");
        }
        else if( pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == iCount )
        {
            DeleteLast();
        }
        else
        {   
            int iCnt = 0;
            node temp = first;
            node target = null;

            for(iCnt = 1;iCnt < pos - 1 ; iCnt++)
            {
                temp = temp.next;
            }
            target = temp.next;

            temp.next = target.next;
            target.next.prev = temp;
            
            iCount--;

        }
    }



}

class program378
{
 public static void main(String A[])
    {
        DoublyCircularLL sobj = new DoublyCircularLL();

        int iRet = 0;

        sobj.InsertFirst(51);
        sobj.InsertFirst(21);
        sobj.InsertFirst(11);

        sobj.InsertLast(101);
        sobj.InsertLast(111);
        sobj.InsertLast(121);

        sobj.Display();

        iRet = sobj.Count();

        System.out.println("Number of elements in the linked-list are: "+iRet);


        sobj.InsertAtPos(105,5);

        sobj.Display();

        iRet = sobj.Count();

        System.out.println("Number of elements in the linked-list are: "+iRet);

        sobj.DeleteAtPos(5);

        sobj.Display();

        iRet = sobj.Count();

        System.out.println("Number of elements in the linked-list are: "+iRet);
    }
}
